📖 Document basic terms used by the AtlasEngine stack (#4)
atlas-engine/ContributionGuidelines: docs/getting_started/basic_terms.md

# Basic terms

Below you will find a definition for all basic terms we use through the AtlasEngine stack.

## Process Model

A Process Model describes a single BPMN process you have modelled.
It usually consists of at least one lane, a Start Event, a set of tasks and an End Event.

> For example:

Reviewer: ```suggestion
For example:

```

![Process Model](./images/process_model.png)

In this case, the name of the Process Model is `check_age_of_buyer`.

> It has a lane named `Broker` and uses a Start Event, a UserTask, an Exclusive Gateway and two End Events.

Reviewer: ```suggestion
It has a Lane named `Broker` and uses a Start Event, a User Task, an Exclusive Gateway and two End Events.
```

A Process Definition is a collection of multiple Process Models contained in the same `.bpmn` file.

You can store as many Process Models within a Process Definition as you like.

Usually, you'd want to do this, when you want to illustrate how multiple processes are communicating with each other.

For example:

![Process Definition](./images/process_definition.png)

## Process Instance

A Process Instance is created, whenever you **execute** a Process Model.

Each Process Instance has a unique ID that you can use to track its progress, or to retrieve it from the AtlasEngine later.

In BPMN Studio, executing a Process Instance can be viewed with the `Live Execution Tracker`. It will look something like this:

![Process Instance](./images/process_instance.png)

## Correlation

A `Correlation` can be used to group multiple Process Instances together.
They are simple strings that can be attached to any ProcessInstance.

Correlations are useful for applying a common context to a group of Process Instances.

By default, the AtlasEngine will auto-generate a Correlation ID for each Process Instance, but you can also pass your own with each `StartProcessInstance` request.

> For example, when you have a backup process that runs once every week and you wish to group them by the month in which they were executed, you can give them all the correlation id `june_backup`.

Reviewer: ```suggestion
For example, when you have a backup process that runs once every week and you wish to group them by the month in which they were executed, you can give them all the Correlation ID `june_backup`.
```

## Flow Node

A Flow Node is a single executable element on a Process Model.

This includes any kind of event, task and gateway.

Refer to the [BPMN Elements](bpmn_elements.md) section for information about all Flow Nodes that are currently supported.

> 

Reviewer: ```suggestion
```

## Flow Node Instance

A Flow Node Instance is created for each Flow Node that gets executed on a Process Instance.

Taking the same sample from above:

![Process Instance](./images/process_instance.png)

We'd have two Flow Node Instances already existing for this Process Instance:
- One for the Start Event, which is already finished
- One for the Service Task, which is still in progress

> Another Flow Node Instance will be created for the End event, as soon as the Service Task has finished execution.

Reviewer: ```suggestion
Another Flow Node Instance will be created for the End Event, as soon as the Service Task has finished execution.
```

## Process Token

When a Process Instance is executed, it progresses through the different elements of the diagram and captures states for each Flow Node Instance along the way.

This progression and state accumulation can be visualized as a token moving along the Process Instance's path in the diagram.

### Tokens generated by a Flow Node Instance

Each Flow Node Instance typically creates the following tokens:
- `onEnter`: When the Flow Node Instance begins execution

> - `onExit`: When the Flow Node Instance finishes execution

Reviewer: Manchmal wird ein `:` verwendet bei solchen Aufzählungen und manchmal `-`. Bitte vereinheitlichen. 

Genauso wie das Setzen eines Punkts am Ende von Aufzählungen. Manchmal ist es vorhanden und manchmal nicht.

> Flow Node Instances that require User Input or require a specific event to take place before they can continue, also generate additional tokens:

Reviewer: ```suggestion
Flow Node Instances that require user input or require a specific event to take place before they can continue, also generate additional tokens:
```

Reviewer: Hier fehlt übrigens der `onError` token der auftreten kann

Author: Nope, die Token folgen alle dem Namens Schema `onEnter`, `onSuspend`, `onResume` und `onExit` (Anfang, Pause, Fortsetzen, Ende). 

Error, Terminated usw, wird nur über den State abgebildet.

- `onSuspend`: When the Flow Node Instance pauses execution to wait for the continuing event to occur
- `onResume`: When the Flow Node Instance resumed, after the continuing event has occured

### How to use the Process token

> You can use data from the Process Token in any number of ways.

Reviewer: ```suggestion
You can use data from the Process Token in any number of ways.

```

# Basic terms

Below you will find a definition for all basic terms we use through the AtlasEngine stack.

## Process Model

A Process Model describes a single BPMN process you have modelled.
It usually consists of at least one lane, a Start Event, a set of tasks and an End Event.

For example:

> ![Process Model](./images/process_model.png)

Reviewer 2: Hast du das Diagram noch? Wenn ja würde ich anstatt eine png ein svg verwenden (kannst du ja per BPMN Studio erstellen). Da die Qualität dadurch deutlich beser wird.

Author: Nope hab ich nicht.
Merk ich mir mal für die Zukunft.

In this case, the name of the Process Model is `check_age_of_buyer`.
It has a Lane named `Broker` and uses a Start Event, a User Task, an Exclusive Gateway and two End Events.

## Process Definition

A Process Definition is a collection of multiple Process Models contained in the same `.bpmn` file.

You can store as many Process Models within a Process Definition as you like.

Usually, you'd want to do this, when you want to illustrate how multiple processes are communicating with each other.

For example:

> ![Process Definition](./images/process_definition.png)

Reviewer 2: s.o.
